### PR TITLE
Add BERT with fine tuning

### PR DESCRIPTION
**Pull request by Naqu6**
Work done:
- Moved `EncoderModel` into an abstract class then implemented two classes that extend that, one for training encoders with frozen language models and another for training encoders and fine-tuning language models
- Added code to `caching` module to handle caching encoded text (input to a language model)
- Added code to `encoder` datasets module to handle building datasets for fine tuning language models
- Added code to `train/language model` module to handle generating and caching encodings
- Modified the `BaseLanguageModel` class to have better code reuse
- Modified the `GPT-1` language model wrapper implementation
- Added language model wrapper implementations for BERT, RoBERTa
- Updated unit tests for modeling
- Updated tests for encoder datasets
- Updated tests for caching module

Effect on Metrics:
R@1: 10% -> 13%
Median Rank: 16 -> 12

### CODE REVIEW

File: videoretrieval/languagemodels/gpt.py [hunk 1]
```diff
@@ -16,75 +16,31 @@
...
+        tokenizer = OpenAIGPTTokenizerFast.from_pretrained("openai-gpt")
+        tokenizer.pad_token = " "
+        return tokenizer
+
+    def get_model(self):
+        return TFOpenAIGPTModel.from_pretrained("openai-gpt")
 
     @property
     def contextual_embeddings_shape(self):
         return (37, 768)
```
Comment (ozdemirbahadir) [inline videoretrieval/languagemodels/gpt.py:46]: Can you comment on what these numbers correspond to?

A better way is to assign them to variables named carefully and returning those variables. Then, you won't need any comment.

Comment (sbucak) [reply videoretrieval/languagemodels/gpt.py:46]: +1

File: videoretrieval/cache/language_model_cache.py [hunk 2]
```diff
@@ -21,47 +21,73 @@
...
+encodings_schema = {
+    "video_id": tf.io.FixedLenFeature([], tf.string),
+    "serialized_encodings": tf.io.FixedLenFeature([], tf.string),
+    "serialized_attention_masks": tf.io.FixedLenFeature([], tf.string)
+}
 
-def get_feature(value):
+#base_path = Path(f"/mnt/disks/fast_ssd/cached_data/")
+base_path = Path("./cached_data/")
+def get_bytes_feature(value):
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/cache/language_model_cache.py:39]: 👍

File: videoretrieval/cache/language_model_cache.py [hunk 3]
```diff
@@ -21,47 +21,73 @@
...
-base_path = Path(f"/mnt/disks/fast_ssd/cached_data/")
-
+encodings_schema = {
+    "video_id": tf.io.FixedLenFeature([], tf.string),
+    "serialized_encodings": tf.io.FixedLenFeature([], tf.string),
+    "serialized_attention_masks": tf.io.FixedLenFeature([], tf.string)
+}
 
-def get_feature(value):
+#base_path = Path(f"/mnt/disks/fast_ssd/cached_data/")
```
[the remark this answers is not in the document]
Comment (Naqu6) [reply videoretrieval/cache/language_model_cache.py:37]: 👍

File: videoretrieval/languagemodels/gpt.py (hunk 1 above)
Comment (Naqu6) [reply videoretrieval/languagemodels/gpt.py:46]: Updated!

**Pull request merged**